python/hail2/expr/expression.py

```diff
@@ -0,0 +1,1175 @@
+        raise ValueError("Cannot implicitly capture value `{}' with type `{}'.".format(e, e.__class__))
+
+
+@decorator
+def args_to_expr(func, *args):
+    return func(*(to_expr(a) for a in args))
+
+
+def compose(*exprs):
+    src = None
+    indices = set()
+    agg = ()
+    for e in exprs:
+        assert isinstance(e, Expression) or isinstance(e, Aggregable), e
+        # c = to_expr(c)
+        csrc = e._source
+        if not src:
+            src = csrc
+        elif csrc and csrc is not src:
+            raise ValueError('Cannot compose operations between {} and {}'.format(repr(src), repr(csrc)))
+        intersection = indices.intersection(e._indices)
+        left = indices - intersection
+        right = e._indices - intersection
+        if right:
+            for i in left:
+                print('broadcasting index {} along axes [{}]'.format(i, ', '.join(right)))
+        if left:
+            for i in right:
+                print('broadcasting index {} along axes [{}]'.format(i, ', '.join(left)))
+        indices = indices.union(e._indices)
+        agg += e._aggregations
+    return indices, src, agg
+
+
+def convert_expr(x):
+    if isinstance(x, Expression) and x._type.__class__ in typ_to_column:
+        x = typ_to_column[x._type.__class__](x._ast, x._type, x._indices, x._source, x._aggregations)
+
+        if isinstance(x, ArrayExpression) and x._type.element_type.__class__ in elt_typ_to_array_column:
+            return elt_typ_to_array_column[x._type.element_type.__class__](x._ast, x._type, x._indices, x._source,
+                                                                           x._aggregations)
+        elif isinstance(x, SetExpression) and x._type.element_type.__class__ in elt_typ_to_set_column:
+            return elt_typ_to_set_column[x._type.element_type.__class__](x._ast, x._type, x._indices, x._source,
+                                                                         x._aggregations)
+        else:
+            return x
+    else:
+        raise NotImplementedError("Can't convert column with type `" + str(x._type.__class__) + "'.")
+
+__numeric_types = [TInt32, TInt64, TFloat32, TFloat64]
+
+@typecheck(t=Type)
+def is_numeric(t):
+    return t.__class__ in __numeric_types
+
+@typecheck(types=tupleof(Type))
+def convert_numeric_typ(*types):
+    priority_map = {t: p for t, p in zip(__numeric_types, range(len(__numeric_types)))}
+    priority = 0
+    for t in types:
+        assert (is_numeric(t)), t
+        t_priority = priority_map[t.__class__]
+        if t_priority > priority:
+            priority = t_priority
+    return __numeric_types[priority]()
+
+
+class Expression(object):
+    @typecheck_method(ast=AST, type=Type, indices=setof(strlike), source=anytype, aggregations=tupleof(anytype))
+    def __init__(self, ast, type, indices=set(), source=None, aggregations=()):
+        self._ast = ast
+        self._type = type
+        self._indices = indices
+        self._source = source
```

Comment:
I don't quite buy that `_source` makes sense in light of joins.  Shouldn't each index have its own source?

Reply:
After discussion, this seems OK.